File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/settings/SettingsFragment.java
package forpdateam.ru.forpda.ui.fragments.settings;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.support.v7.preference.Preference;
import android.view.View;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.Toast;

import forpdateam.ru.forpda.App;
import forpdateam.ru.forpda.BuildConfig;
import forpdateam.ru.forpda.R;
import forpdateam.ru.forpda.apirx.RxApi;
import forpdateam.ru.forpda.client.ClientHelper;
import forpdateam.ru.forpda.common.IntentHandler;
import forpdateam.ru.forpda.common.Preferences;
import forpdateam.ru.forpda.ui.activities.SettingsActivity;
import forpdateam.ru.forpda.ui.activities.updatechecker.UpdateCheckerActivity;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by radiationx on 25.12.16.
 */

public class SettingsFragment extends BaseSettingFragment {
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        addPreferencesFromResource(R.xml.preferences);
        if (ClientHelper.getAuthState()) {
            findPreference("auth.action.logout")
                    .setOnPreferenceClickListener(preference -> {
                        new AlertDialog.Builder(getActivity())
                                .setMessage(R.string.ask_logout)
                                .setPositiveButton(R.string.ok, (dialog, which) -> RxApi.Auth().logout().onErrorReturn(throwable -> false)
                                        .subscribeOn(Schedulers.io())
                                        .observeOn(AndroidSchedulers.mainThread())
                                        .subscribe(aBoolean -> {
                                            if (aBoolean) {
                                                Toast.makeText(App.getContext(), "Logout complete", Toast.LENGTH_LONG).show();
                                                ClientHelper.get().notifyAuthChanged(ClientHelper.AUTH_STATE_LOGOUT);
                                            } else {
                                                Toast.makeText(App.getContext(), "Logout error", Toast.LENGTH_LONG).show();
                                            }
                                        }))
                                .setNegativeButton(R.string.no, null)
                                .show();
                        return false;
                    });
        } else {
            findPreference("auth.action.logout")
                    .setEnabled(false);
        }

        findPreference("about.application")
                .setSummary(String.format(getString(R.string.version_Build), BuildConfig.VERSION_NAME));

        {
            Preference fPref = findPreference("about.app_faq");
            fPref.setIcon(App.getAppVecDrawable(R.drawable.contact_site));
            fPref.setOnPreferenceClickListener(preference -> {
                IntentHandler.externalIntent("http://4pda.ru/forum/index.php?s=&showtopic=820313&view=findpost&p=64077514");
                return false;
            });
        }

        {
            Preference fPref = findPreference("about.app_topic");
            fPref.setIcon(App.getAppVecDrawable(R.drawable.contact_site));
            fPref.setOnPreferenceClickListener(preference -> {
                IntentHandler.externalIntent("https://4pda.ru/forum/index.php?showtopic=820313");
                return false;
            });
        }

        findPreference("about.check_update")
                .setOnPreferenceClickListener(preference -> {
                    startActivity(new Intent(getActivity(), UpdateCheckerActivity.class));
                    return false;
                });

        findPreference(Preferences.Main.WEBVIEW_FONT_SIZE)
                .setOnPreferenceClickListener(preference -> {
                    View v = getActivity().getLayoutInflater().inflate(R.layout.dialog_font_size, null);

                    assert v != null;
                    final SeekBar seekBar = (SeekBar) v.findViewById(R.id.value_seekbar);
                    seekBar.setProgress(Preferences.Main.getWebViewSize(getContext()) - 1 - 7);
                    final TextView textView = (TextView) v.findViewById(R.id.value_textview);
                    textView.setText(Integer.toString(seekBar.getProgress() + 1 + 7));
                    textView.setTextSize(seekBar.getProgress() + 1 + 7);
                    seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
                        @Override
                        public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                            textView.setText(Integer.toString(i + 1 + 7));
                            textView.setTextSize(i + 1 + 7);
                        }

                        @Override
                        public void onStartTrackingTouch(SeekBar seekBar) {

                        }

                        @Override
                        public void onStopTrackingTouch(SeekBar seekBar) {

                        }
                    });
                    AlertDialog dialog = new AlertDialog.Builder(getActivity())
                            .setTitle(R.string.text_size)
                            .setView(v)
                            .setPositiveButton(R.string.ok, (dialog1, which) -> Preferences.Main.setWebViewSize(getContext(), seekBar.getProgress() + 1 + 7))
                            .setNegativeButton(R.string.cancel, null)
                            .setNeutralButton(R.string.reset, null)
                            .show();
                    dialog.getButton(DialogInterface.BUTTON_NEUTRAL).setOnClickListener(v1 -> {
                        seekBar.setProgress(16 - 1 - 7);
                        Preferences.Main.setWebViewSize(getContext(), 16);
                        App.get().getPreferences().edit().putInt(Preferences.Main.WEBVIEW_FONT_SIZE, 16).apply();
                    });
                    return false;
                });

        findPreference("open_notifications")
                .setOnPreferenceClickListener(preference -> {
                    Intent intent = new Intent(getActivity(), SettingsActivity.class);
                    intent.putExtra(SettingsActivity.ARG_NEW_PREFERENCE_SCREEN, NotificationsSettingsFragment.PREFERENCE_SCREEN_NAME);
                    startActivity(intent);
                    return true;
                });
    }
}

File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/settings/SettingsFragment.kt
package forpdateam.ru.forpda.ui.fragments.settings

import android.annotation.SuppressLint
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.SeekBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.preference.Preference
import forpdateam.ru.forpda.App
import forpdateam.ru.forpda.R
import forpdateam.ru.forpda.common.AppBuildConfig
import forpdateam.ru.forpda.common.Preferences
import forpdateam.ru.forpda.extensions.coRunCatching
import forpdateam.ru.forpda.ui.activities.SettingsActivity
import forpdateam.ru.forpda.ui.activities.updatechecker.UpdateCheckerActivity
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

/**
 * Created by radiationx on 25.12.16.
 */

class SettingsFragment : BaseSettingFragment() {
    private val authRepository = App.get().Di().authRepository
    private val authHolder = App.get().Di().authHolder
    private val mainPreferencesHolder = App.get().Di().mainPreferencesHolder
    private var logoutJob: Job? = null

    @SuppressLint("InflateParams")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        addPreferencesFromResource(R.xml.preferences)
        if (authHolder.get().isAuth()) {
            findPreference<Preference>("auth.action.logout")?.apply {
                setOnPreferenceClickListener {
                    AlertDialog.Builder(requireActivity())
                        .setMessage(R.string.ask_logout)
                        .setPositiveButton(R.string.ok) { _, _ ->
                            logoutRequest()
                        }
                        .setNegativeButton(R.string.no, null)
                        .show()
                    false
                }
            }
        } else {
            findPreference<Preference>("auth.action.logout")?.apply {
                isEnabled = false
            }
        }

        findPreference<Preference>("clear_menu_sequence")?.apply {
            setOnPreferenceClickListener {
                AlertDialog.Builder(requireActivity())
                    .setMessage("Подтвердите действие")
                    .setPositiveButton(R.string.ok) { _, _ ->
                        App.get().Di().preferences.edit().remove("menu_items_sequence").apply()
                    }
                    .setNegativeButton(R.string.cancel, null)
                    .show()
                false
            }
        }

        findPreference<Preference>("about.application")?.apply {
            summary = String.format(getString(R.string.version_Build), AppBuildConfig.versionName)
        }

        findPreference<Preference>("about.check_update")?.apply {
            setOnPreferenceClickListener {
                startActivity(Intent(activity, UpdateCheckerActivity::class.java))
                false
            }
        }

        findPreference<Preference>(Preferences.Main.WEBVIEW_FONT_SIZE)?.apply {
            setOnPreferenceClickListener { _ ->

                val dialogView =
                    requireActivity().layoutInflater.inflate(R.layout.dialog_font_size, null)!!
                val seekBar = dialogView.findViewById<View>(R.id.value_seekbar) as SeekBar
                val textView = dialogView.findViewById<View>(R.id.value_textview) as TextView

                seekBar.progress = mainPreferencesHolder.webViewFontSize.get() - 1 - 7

                textView.text = (seekBar.progress + 1 + 7).toString()
                textView.textSize = (seekBar.progress + 1 + 7).toFloat()

                seekBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
                    override fun onProgressChanged(seekBar: SeekBar, i: Int, b: Boolean) {
                        textView.text = (i + 1 + 7).toString()
                        textView.textSize = (i + 1 + 7).toFloat()
                    }

                    override fun onStartTrackingTouch(seekBar: SeekBar) {}
                    override fun onStopTrackingTouch(seekBar: SeekBar) {}
                })
                AlertDialog.Builder(requireActivity())
                    .setTitle(R.string.text_size)
                    .setView(dialogView)
                    .setPositiveButton(R.string.ok) { _, _ ->
                        mainPreferencesHolder.webViewFontSize.set(seekBar.progress + 1 + 7)
                    }
                    .setNegativeButton(R.string.cancel, null)
                    .setNeutralButton(R.string.reset, null)
                    .show()
                    .getButton(DialogInterface.BUTTON_NEUTRAL)
                    .setOnClickListener {
                        seekBar.progress = 16 - 1 - 7
                        mainPreferencesHolder.webViewFontSize.set(16)
                    }

                false
            }
        }

        findPreference<Preference>("open_notifications")?.apply {
            setOnPreferenceClickListener {
                val intent = Intent(activity, SettingsActivity::class.java)
                intent.putExtra(
                    SettingsActivity.ARG_NEW_PREFERENCE_SCREEN,
                    NotificationsSettingsFragment.PREFERENCE_SCREEN_NAME
                )
                startActivity(intent)
                true
            }
        }
    }

    private fun logoutRequest() {
        logoutJob?.cancel()
        logoutJob = GlobalScope.launch(Dispatchers.Main) {
            coRunCatching {
                authRepository.signOut()
            }.onSuccess {
                Toast.makeText(App.getContext(), "Logout complete", Toast.LENGTH_LONG).show()
            }.onFailure {
                Toast.makeText(App.getContext(), "Logout error: $it", Toast.LENGTH_LONG).show()
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        logoutJob?.cancel()
    }
}
